src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNetIntegration.cs

```diff
@@ -72,44 +93,77 @@ public static object ExecuteDbDataReader(object @this, int behavior, int opCode,
-            TargetType = "System.Data.Common.DbCommand",
-            TargetSignatureTypes = new[] { "System.Threading.Tasks.Task`1<System.Data.Common.DbDataReader>", "System.Data.CommandBehavior", "System.Threading.CancellationToken" },
+            TargetAssembly = FrameworkAssembly,
+            TargetType = DbCommand,
+            TargetSignatureTypes = new[] { "System.Threading.Tasks.Task`1<System.Data.Common.DbDataReader>", CommandBehavior, ClrNames.CancellationToken },
             TargetMinimumVersion = Major4,
             TargetMaximumVersion = Major4)]
         [InterceptMethod(
-            TargetAssembly = "System.Data.Common", // .NET Core
-            TargetType = "System.Data.Common.DbCommand",
-            TargetSignatureTypes = new[] { "System.Threading.Tasks.Task`1<System.Data.Common.DbDataReader>", "System.Data.CommandBehavior", "System.Threading.CancellationToken" },
+            TargetAssembly = CoreAssembly,
+            TargetType = DbCommand,
+            TargetSignatureTypes = new[] { "System.Threading.Tasks.Task`1<System.Data.Common.DbDataReader>", CommandBehavior, ClrNames.CancellationToken },
             TargetMinimumVersion = Major4,
             TargetMaximumVersion = Major4)]
-        public static object ExecuteDbDataReaderAsync(object @this, int behavior, object cancellationTokenSource, int opCode, int mdToken)
+        public static object ExecuteDbDataReaderAsync(
+            object @this,
+            int behavior,
+            object cancellationTokenSource,
+            int opCode,
+            int mdToken)
         {
             var tokenSource = cancellationTokenSource as CancellationTokenSource;
             var cancellationToken = tokenSource?.Token ?? CancellationToken.None;
 
             var command = (DbCommand)@this;
+            var instanceType = command.GetType();
             var commandBehavior = (CommandBehavior)behavior;
+            var callingAssembly = Assembly.GetCallingAssembly();
+            var dataReaderType = callingAssembly.GetType(DbDataReader);
+
+            Func<object, object, object, object> instrumentedMethod = null;
 
-            return ExecuteDbDataReaderAsyncInternal(command, commandBehavior, cancellationToken);
+            try
+            {
+                instrumentedMethod =
+                    MethodBuilder<Func<object, object, object, object>>
+                       .Start(Assembly.GetCallingAssembly(), mdToken, opCode, nameof(ExecuteDbDataReaderAsync))
+                       .WithConcreteType(instanceType)
+                       .WithParameters(commandBehavior, cancellationToken)
+                       .Build();
+            }
+            catch (Exception ex)
+            {
+                Log.ErrorException($"Error resolving {DbCommand}.{nameof(ExecuteDbDataReaderAsync)}(...)", ex);
+                throw;
+            }
+
+            return AsyncHelper.InvokeGenericTaskDelegate(
+                instanceType,
```

Comment:
Should the owning type be the instance type? I thought the owning type would be the abstract DbCommand type, like how we work with interfaces

Reply:
The code above is a bit unclear:
```
var command = (DbCommand)@this;
var instanceType = command.GetType();
```

Reply:
I'm unsure on whether there should be a specific pattern for this. Are there pros or cons to going either way, or insisting on one way? e.g., only passing the (maybe base or abstract) type we're looking for; only passing the concrete type

Reply:
I just realized the InvokeGenericTaskDelegate only uses the owningType to create a key in the method cache, so using the interface type as the owning type could reduce the number of times we create closed generic versions of the given open generic method and also reduce the number of entries in the cache. We don't have performance measurements right now so we can't really quantify the difference, but it is much easier to grab the concrete type so I think it's fine as written.